aten/src/ATen/native/quantized/cpu/qupsample_nearest3d.cpp

```diff
@@ -0,0 +1,200 @@
+    const int64_t d1 =
+          nearest_neighbor_compute_source_index(depth_scale, d2, input_depth);
+
+    for (int64_t h2 = 0; h2 < output_height; ++h2) {
+      const int64_t h1 =
+          nearest_neighbor_compute_source_index(height_scale, h2, input_height);
+
+      for (int64_t w2 = 0; w2 < output_width; ++w2) {
+        const int64_t w1 =
+            nearest_neighbor_compute_source_index(width_scale, w2, input_width);
+
+        const auto* pos1 = &i_p[d1 * input_height * input_width + h1 * input_width + w1];
+        auto* pos2 = &o_p[d2 * output_height * output_width + h2 * output_width + w2];
+
+        for (int64_t c = 0; c < channels; ++c) {
+          pos2[0] = pos1[0];
+          pos1 += input_depth * input_height * input_width;
+          pos2 += output_depth * output_height * output_width;
+        }
+      }
+    }
+  }
+}
+
+template <typename scalar_t>
+static void upsample_nearest3d_out_frame_nhwc(
+    scalar_t* odata,
+    scalar_t* idata,
+    int64_t input_depth,
+    int64_t input_height,
+    int64_t input_width,
+    int64_t output_depth,
+    int64_t output_height,
+    int64_t output_width,
+    int64_t nbatch,
+    int64_t channels,
+    c10::optional<double> scales_d,
+    c10::optional<double> scales_h,
+    c10::optional<double> scales_w) {
+  float depth_scale = compute_scales_value<float>(scales_d, input_depth, output_depth);
+  float height_scale = compute_scales_value<float>(scales_h, input_height, output_height);
+  float width_scale = compute_scales_value<float>(scales_w, input_width, output_width);
+
+  for (int b = 0; b < nbatch; b++) {
+    auto* i_p = reinterpret_cast<typename scalar_t::underlying*>(idata + b * input_depth * input_height * input_width * channels);
+    auto* o_p = reinterpret_cast<typename scalar_t::underlying*>(odata + b * output_depth * output_height * output_width * channels);
+    // special case: just copy
+    if (input_depth == output_depth && input_height == output_height && input_width == output_width) {
+      std::memcpy(o_p, i_p, channels * input_depth * input_height * input_width * sizeof(typename scalar_t::underlying));
+      return;
+    }
+
+    for (int64_t d2 = 0; d2 < output_depth; ++d2) {
+      const int64_t d1 =
+          nearest_neighbor_compute_source_index(depth_scale, d2, input_depth);
+      for (int64_t h2 = 0; h2 < output_height; ++h2) {
+        const int64_t h1 =
+            nearest_neighbor_compute_source_index(height_scale, h2, input_height);
+
+        for (int64_t w2 = 0; w2 < output_width; ++w2) {
+          const int64_t w1 =
+              nearest_neighbor_compute_source_index(width_scale, w2, input_width);
+
+          const auto* pos1 = &i_p[(d1 * input_height * input_width + h1 * input_width + w1)*channels];
+          auto* pos2 = &o_p[(d2 * output_height * output_width + h2 * output_width + w2)*channels];
+          std::memcpy(pos2, pos1, channels * sizeof(typename scalar_t::underlying));
```

Comment:
When channels is very small the `memcpy` dispatch can end up being very expensive, just something to watch out for